src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -1647,28 +1806,75 @@ - (void)_handleReportEnd
 
 // For unit testing only
 #ifdef DEBUG
-    id delegate = _weakDelegate.strongObject;
-    if (delegate) {
-        dispatch_async(_delegateQueue, ^{
-            if ([delegate respondsToSelector:@selector(unitTestReportEndForDevice:)]) {
-                [delegate unitTestReportEndForDevice:self];
+    [self _callDelegatesWithBlock:^(id testDelegate) {
+        if ([testDelegate respondsToSelector:@selector(unitTestReportEndForDevice:)]) {
+            [testDelegate unitTestReportEndForDevice:self];
+        }
+    }];
+#endif
+}
+
+- (BOOL)_interestedPaths:(NSArray * _Nullable)interestedPaths includesPath:(MTRAttributePath *)attributePath
+{
+    for (id interestedPath in interestedPaths) {
+        if ([interestedPath isKindOfClass:[MTRClusterPath class]]) {
+            MTRClusterPath * interestedClusterPath = interestedPath;
+            if (interestedClusterPath.cluster.unsignedLongValue == attributePath.cluster.unsignedLongValue) {
+                return YES;
             }
-        });
+        } else if ([interestedPath isKindOfClass:[MTRAttributePath class]]) {
+            MTRAttributePath * interestedAttributePath = interestedPath;
+            if ((interestedAttributePath.cluster.unsignedLongValue == attributePath.cluster.unsignedLongValue) && (interestedAttributePath.attribute.unsignedLongValue == attributePath.attribute.unsignedLongValue)) {
+                return YES;
+            }
+        }
     }
-#endif
+
+    return NO;
+}
+
+// Returns filtered set of attributes using an interestedPaths array.
+// Returns nil if no attribute reports has a path that matches the paths in the interestedPaths array.
+- (NSArray<NSDictionary<NSString *, id> *> *)_filteredAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes forInterestedPaths:(NSArray * _Nullable)interestedPaths
+{
+    if (!interestedPaths.count) {
+        return attributes;
+    }
+
+    NSMutableArray * filteredAttributes = nil;
+    for (NSDictionary<NSString *, id> * responseValue in attributes) {
+        MTRAttributePath * attributePath = responseValue[MTRAttributePathKey];
+        if ([self _interestedPaths:interestedPaths includesPath:attributePath]) {
```

Comment:
A question.  Should `_interestedPaths:includesPath:` be a class method?  It does not use any "self" state....  Not sure whether that's a convention ObjC uses.

Or even a static function or something.

Reply:
Static methods are a little tedious to use in ObjC because you have to explicitly call them with the class (or `self.class`), so I think it's fine for helpers to not use state and still be instance methods. static functions are OK can feel inconsistent due to the different style signatures. I think we should just put `MTR_DIRECT_MEMBERS` on all `@implementation` blocks, then these internal helpers effectively become static C functions but with ObjC syntax.